nixos: nixos-generate-config: document all the common things

oxij opened the pull request:
###### Motivation for this change

Combination of https://github.com/NixOS/nixpkgs/issues/17126#issuecomment-315190218 and https://github.com/NixOS/nixpkgs/pull/35292#issuecomment-367714173 got me over the edge.

###### Things done

- [X] It evaluates.

Feel free to point to typos I probably missed.

Mic92 commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,36 +638,178 @@ sub multiLineList {
+  # To disable the firewall altogether use the following.
+  #
+  # Use only inside trusted networks behind IPv4 NAT and without IPv6
+  # support or else bad people across the galaxy will be able to
```
Only because the network has ipv6, does not mean your router does not block incoming connections to you IP.
And also NAT is not a firewall.

Suggestion:

```
Disable your firewall only in trusted networks with inbound firewalls or if your services are supposed to public on the internet.
```

Mic92 commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,36 +638,178 @@ sub multiLineList {
+  # (Actually, not enabling this doesn't disable ALSA because ALSA is
+  # the Linux kernel default. But you won't get access to ALSA tools
+  # like alsamixer and your ALSA settings like volume won't get saved
+  # and restored between reboots without enabling the above.)
```
This can be shortened:

```
This option will add alsatools to your system and make ALSA settings such as volumes persistent across reboots.
```

oxij commented:
Both nitpicks fixed.

mogorman commented:
this file is getting huge, but i do think several the tips you recommend are good / useful

oxij commented:
> this file is getting huge

It is, but it now documents literally everything you commonly need to know about configuring your system after reading the first 12 items of the installation chapter of the manual. All of my `configuration.nix` files (except those that I use for tests) touch all of the mentioned variables. I assume that most people read first 12 items and then start editing `configuration.nix`, so it is the most appropriate place to put all of that.

The part about `nix-env` and `allowUnfree` should still be copied to the manual (actually, to both `nix` and `nixos` manuals), but that's out of scope of this PR.

oxij commented:
Fixed a tiny typo/OCD thing.

What's holding this? While this ferments some people encounter the problem described in #17126, say "f*ck it" and drop NixOS while others unwittingly enable the PA daemon.

Unless there's a problem with this I would merge this ASAP and cherry-pick to 18.03 too.

/cc @grahamc @vcunat

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,36 +638,178 @@ sub multiLineList {
+  # You might want to enable this if you run a minimalistic desktop
+  # environment or work from bare linux ttys/framebuffers.
+
+  # sound.mediaKeys.enable = true;
```
I think removing this section about mediakeys would make it easier to merge, since mediakeys being an option is a bit contentious.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,36 +638,178 @@ sub multiLineList {
+  # (You might need to merge this line with the above mention of
+  # environment.systemPackages as you can't assign the same attribute
+  # multiple times in a single nix attribute set.)
+
```
Should probably drop this previous section, because the section about enabling sound explains it well enough I think.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -564,10 +564,14 @@ sub multiLineList {
+  # networking.hostName = "nixos";
+
+  # Enable wireless support via wpa_supplicant.
+  # networking.wireless.enable = true;
```
:+1:

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -564,10 +564,14 @@ sub multiLineList {
+  # networking.wireless.enable = true;
 
   # Select internationalisation properties.
+
```
I think these newlines confuse it, where without the newline it is clearly tied to the section.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +581,53 @@ sub multiLineList {
   #   wget vim
   # ];
 
+  # To search by name, run:
```
"To search packages"

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +581,53 @@ sub multiLineList {
 
-  # List packages installed in system profile. To search by name, run:
-  # \$ nix-env -qaP | grep wget
+  # List packages installed in system profile.
```
List of packages for the system profile, by attribute name.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +581,53 @@ sub multiLineList {
 
+  # To search by name, run:
+  #
+  # \$ nix-env -qaP | grep wget
```
`-qaP '.*wget.*'`

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +581,53 @@ sub multiLineList {
+  #
+  # into your ~/.config/nixpkgs/config.nix file.
+  #
+  # Finally, note that if you don't use nixos channels, you can still use
```
the NixOS channels

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +581,53 @@ sub multiLineList {
+  # Finally, note that if you don't use nixos channels, you can still use
+  # nix-env by running it directly from the nixpkgs tree like so
+  #
+  # \$ nix-env -f ./default.nix -qaP | grep wget
```
`-qaP '.*wget.*'`

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +581,53 @@ sub multiLineList {
   # Some programs need SUID wrappers, can be configured further or are
-  # started in user sessions.
+  # started in user sessions. Those usually can be configured with
+  # options under programs.<name> subtree.
```
Maybe add `See man configuration.nix.`

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,36 +638,178 @@ sub multiLineList {
+  # Use only inside trusted networks behind another firewall or else
+  # bad people across the galaxy will be able to access any ports you
+  # leave open.
+  #
```
Probably a bit much :)

grahamc replied:
I didn't say it was completely worthless, just a bit much.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,36 +638,178 @@ sub multiLineList {
 
   # Enable touchpad support.
   # services.xserver.libinput.enable = true;
 
```
I think all of the changes following here should probably be in the wiki instead.

grahamc reviewed (commented):
I like a good bit of it and would merge most of it.

oxij commented:
> I think these newlines confuse it, where without the newline it is clearly tied to the section.

Ok, fixed everywhere. It now groups things with an empty comment line.

> "To search packages"

Fixed.

> List of packages for the system profile, by attribute name.

Fixed.

> `-qaP '.*wget.*'`

Fixed everywhere.

> the NixOS channels

Fixed.

> Maybe add `See man configuration.nix.`

Done.

> Probably a bit much :)

It's not. See https://www.youtube.com/watch?v=yBA6u5IsXyc

> I think removing this section about mediakeys would make it easier to merge, since mediakeys being an option is a bit contentious.

IIRC the contention was "I run X11, my WM does this, I don't want this, please remove it". Since it's not enabled by default I don't think it's a valid argument.

Anyway, I moved that doc to the mediaKeys docstring and made into a tiny note. This way it should make the thing much less contentious, I hope. If you think it's still too much, I'm ok with removing it, but I think it's useful to keep it.

> Should probably drop this previous section, because the section about enabling sound explains it well enough I think.

I disagree, grepping the whole of nixpkgs for `apulse` gives no results that explain how and when to use it except this. Since Mozilla keeps pushing people to PA I think it's only fair to inform people that they can work around it with a single line of config.

Anyway, I rewrote that part to be shorter.

> I think all of the changes following here should probably be in the wiki instead.

I disagree.

I think the default `configuration.nix` should include everything you need to set to reboot into the new configuration and see properly working X11 desktop (including acceleration) with your favorite apps. So allowUnfree, GPU and desktop stuff is essential.

The network user account thing is more contentious, I agree. Note, however, that NixOS already does exactly this for system (as opposed to user) services (hence, another reason to distrust PA, just sayin') and that is not considered contentious for some reason (I suppose, because it requires zero configuration). However, I think it's essential to inform as many users as possible of the easiness of doing things this way. Many security-oriented sites recommend installing software that manages your keys/wallets onto your Android phone (seriously, Android phone!) instead of your desktop because Android does separate users for different apps by default. This makes me mad (because they recommend keeping secrets on a device that almost never updates it's kernel), twice so since they are essentially correct that it still gives more security than your average Linux install (because of the above), trice so since it's super easy to fix on OSes with declarative configs like NixOS. Which is why I wrote this part in the first place.

Anyway, I moved the `stateVersion` thing to the top, and added "you can ignore all of this" note before the thing to make it less contentious. I think it should stay. (Actually, I hope to eventually add notes on QubesOS-like setups there too, when NixOS accepts patches required to do that.)

Also, I think the wiki is either dead or should be dead and everything should be greppable directly in the repo. Also, related, I hate that most code in nixpkgs refers to manual pages via nixos.org URLs instead of just pointing to the text files in the repo itself. This is crazy, but it is understandable, because those files are in unreadable docbook format. The only thing that prevents "you only need to clone nixpkgs and you get everything" utopia is docbook. But that's outside of the scope of this PR.

Anyway, see the new patch. Will squash if all is ok.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -563,6 +563,12 @@ sub multiLineList {
+  # servers. You should change this only after NixOS release notes say you
+  # should.
+  system.stateVersion = "${\(qw(@release@))}"; # Did you read the comment?
+
```
:-1: it shouldn't be near the top if they shouldn't touch it.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +587,64 @@ sub multiLineList {
-  # List packages installed in system profile. To search by name, run:
-  # \$ nix-env -qaP | grep wget
+  # List of packages installed in system profile (by attribute name).
+  #
```
:+1:

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +587,64 @@ sub multiLineList {
+  # still use nix-env by running it directly from the nixpkgs tree like so
+  #
+  # \$ nix-env -f . -qaP '.*wget.*'
+
```
:+1:

edolstra replied:
Given that the whole point of not showing unfree software by default is to not promote the use of unfree software, adding detailed instructions to the default configuration.nix is questionable IMHO.

7c6f434c replied:
@edolstra Well, it is half of the point — the other half is that we have too many different evaluation contexts to ever be sure that `allowUnfree` is not silently true in one of them.

(I do _personally_ agree with the point you make, although I value the fail-safe part more and I am not sure if people stopping opening issues like #17126 is worth documenting this in more places)

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -577,14 +587,64 @@ sub multiLineList {
-  # started in user sessions.
+  # started in user sessions. Those usually can be configured with
+  # options under programs.<name> subtree. See configuration.nix(5)
+  # man page.
```
:+1:

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # across reboots. Not enabling this doesn't actually disable sound
+  # output via ALSA because ALSA is the default sound API of the Linux
+  # kernel. Also see `sound.mediaKeys.enable` in configuration.nix(5).
+  #
```
:+1:

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # Enable sound output via PulseAudio. This option will enable
+  # PulseAudio daemon and configure `asound.conf` to route the
+  # `default` ALSA pcm and ctl to PulseAudio daemon.
+  #
```
:+1:

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # Note that you don't need to enable PulseAudio daemon to get either
+  # of sound playback or software sound mixing. It will out of the box
+  # over ALSA. If you do want to run PulseAudio daemon, however, you
+  # may safely disable `sound.enable`.
```
:+1:

wmertens replied:
How about ``"you may set `sound.enable` to `false`."``?

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # quality to be connected to untrusted data sources like the browser
+  # by default).
+  #
+  # environment.systemPackages = [ pkgs.apulse ];
```
:-1:  but would accept `For more information, see nixos.wiki/something-about-sound...`

7c6f434c replied:
Why refer to a wiki here? I agree with @oxij: this is unlikely to change for a long time.

wmertens replied:
I don't think "some people claim"-type information belongs here. I too would prefer a link to a deeper dive.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # };
+
+  # If you're in a hurry, you may stop editing your configuration.nix now
+  # and completely ignore everything below.
```
:-1: this file will be immediately out of date because nothing can update it. We shouldn't be talking about best practices or general techniques. This belongs in other documentation elsewhere. The parts about unfree software are universally true. The remainder of this should definitely not be in this config file, but instead in the documentation or wiki.

edolstra replied:
I agree, this is the main problem with this PR. It's fine to install a small configuration.nix with some options to get people started, but it should not duplicate the manual or try to cover a gazillion use cases.

7c6f434c replied:
Well, experimental evidence shows that not enough people read the manual anyway…

chreekat replied:
+1. Using a separate user to run e.g. firefox sounds like a great idea, but the boilerplate in configuration.nix is not the soapbox from which to broadcast this position. People approaching this file for the first time just want to get NixOS running, which already has enough mind-blowing paradigm shifts to absorb.

wmertens replied:
agree, :-1: on the somewhat niche and cumbersome desktop-only per-software user approach here. It would be better to a link to general information about running NixOS on the desktop with a section on per-software users.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # services.xserver.displayManager.sddm.enable = true;
+  # services.xserver.desktopManager.plasma5.enable = true;
+
+  # Or maybe you want to enable Xfce Desktop Environment and Slim
```
:-1: on the addition of xfce / slim and the following, they'll need to find the full list of options somehow, and this file isn't in the business of describing every use case. We should be teaching them to fish here.

grahamc commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
-  # services.xserver.displayManager.sddm.enable = true;
-  # services.xserver.desktopManager.plasma5.enable = true;
+  # Enable graphical acceleration for NVIDIA GPUs (requires unfree
+  # packages, see above).
```
:-1: this information is too likely to get out of date, and really can't cover all the important cases. This should lead to external documentation.

grahamc reviewed (commented):
I can't speak to the docbook, but nixos.wiki is a happy and healthy source of pretty high quality information now, thanks to fadenb, lassalus, makefu, mic92, samueldr, and many many others. The risk of putting too much in this file is it is never updated again, making it drift further and further away from reality. I like the idea of putting _more_ information here, but want it to lead to updatable sources.

oxij commented:
> I can't speak to the docbook, but nixos.wiki is a happy and healthy source of pretty high quality information now, thanks to fadenb, lassalus, makefu, mic92, samueldr, and many many others.

While I appreciate the work, I don't like the idea. I think we already had a wiki before and saw what that turned itself into. The problem with wikis tied to software projects is that after initial wave of enthusiasm they get out of sync with the code even faster than the documentation that ships with the code.

Ask yourself. Your last ten nixpkgs problems. How many solutions to them were on the wiki? Somewhere else on the web? In GitHub issues? In the `git log`? I usually start by grepping the repo, the git log, and GitHub issues (I archive and index all of them, so searching them is easy) (in parallel with a simple script), 99.9% of the time I find the solution before the web browser window appears on my screen.

My point is that we should point users to places developers update directly, because that's where all the fish is. Unless nixos.wiki becomes a submodule of nixpkgs I will not touch it. (But if it becomes a submodule then I would ask why can't we just rewrite all nixpkgs documentation in a saner format and just assimilate the wiki up.)

In short, I think documentation related to nixpkgs usage should be in the nixpkgs repo itself, updatable simultaneously with the code, and browsable with just `grep`, `cat` and `git log`.

Arch and Gentoo Linux wikis and StackOverflow are, IMHO, nice examples of how things get outdated fast. Every singe topic I know anything about usually has a bunch of outdated info in the most prominent places of the articles/answers.

Philosophically, what I think is needed here are not wikis, but proper upstream documentation that is compiled into a typechecked hypertext. Bear with me for a second.

Imagine a world where all the documentation is written in `org-mode` format. The beauty of it is that not only it's easy to read and write, but it's also easy to parse (line-based) and it can be compiled to all the common formats like HTML or texinfo (I once wrote `wiki->HTML` converter in three pages `sed`, just for fun, org-mode is even simpler), and it supports references by GUIDs and a set of org-mode files can be checked for broken links automagically.

Now imagine that all upstream packages have proper documentation in said org-mode format. Now, not only can you refer directly to any piece of their documentation, but you can also build your own documentation by linking and including pieces of documentation from upstream projects and that could be compile-time checked.

I.e. imagine NixOS manual having something like

```org
* Setting up sound
** ALSA {{INCLUDESUBTREE|${pkgs.alsaLib}/share/doc/alsa/index.org}}
** JACK {{INCLUDESUBTREE|${pkgs.jackd}/share/doc/jack/index.org}}
** PulseAudio {{INCLUDESUBTREE|${pkgs.pulseaudio}/share/doc/pulseaudio/index.org}}
```

or

```org
* Setting up sound
** ALSA {{INCLUDESUBTREEBYID|ca45533c-1e07-413c-8a87-8b4d1e4cade3}}
** JACK {{INCLUDESUBTREEBYID|...}}
** PulseAudio {{INCLUDESUBTREEBYID|...}}
```

(`{{}}` are macros, nice names for them are up to debate) and the documentation build would then just check those links and fail if something gets out of sync.

But since the human race is not yet ready for this alien technology I prefer to put everything into a single repo and generate all kinds of usable documentation formats from it (not just HTML, see #11499).

(Also note that I didn't mention SLNOS even once in the above :])

> The risk of putting too much in this file is it is never updated again, making it drift further and further away from reality. I like the idea of putting _more_ information here, but want it to lead to updatable sources.

Then how about adding "review `nixos-generate-config`" to the release checklist. You do have release checklist, right?

> I didn't say it was completely worthless, just a bit much. 

Please elaborate what exactly is your problem here. The "galaxy"? How would you write it?

> +  # Note that you don't need to enable PulseAudio daemon to get either
> +  # of sound playback or software sound mixing. It will out of the box
> +  # over ALSA. If you do want to run PulseAudio daemon, however, you
> +  # may safely disable `sound.enable`.
> +
> +  # Also note that you can emulate PulseAudio API over ALSA for apps
> +  # that refuse to work over ALSA themselves by running them under
> +  # apulse. This way you can run one daemon less which benefits your
> +  # security (some people claim that PulseAudio codebase lacks in
> +  # quality to be connected to untrusted data sources like the browser
> +  # by default).
> +  #
> +  # environment.systemPackages = [ pkgs.apulse ];
>
> :-1:  but would accept `For more information, see nixos.wiki/something-about-sound...`

I don't understand which part you disagree with. Do you want me to stop mentioning `apulse`? Do you want me to stop informing users that, contrary to a common misconception, they probably don't even need to run PulseAudio daemon to get all the features they want. Both? (I'm not trolling, I honestly don't understand the argument.)

See above about the wiki.

> :-1: it shouldn't be near the top if they shouldn't touch it.

Ok, will fix.

> :-1: this file will be immediately out of date because nothing can update it.

Eh? "nothing can update it"? What?

If you mean keeping it in sync, then we can add some more markup to the file and the add some nixos tests that would check that different combinations of uncommented options evaluate or whatever.

> We shouldn't be talking about best practices or general techniques. This belongs in other documentation elsewhere. The parts about unfree software are universally true. The remainder of this should definitely not be in this config file, but instead in the documentation or wiki.

See above about the wiki.

However, on this and the rest of your :-1:s, I do see you point. I now see that, indeed, this is not the best place to present a bunch of stuff that gets updated frequently, GPU and DE stuff should be moved to the `configuration.nix(5)` manual and only pointers should be left here.

However, I feel that this argument does not apply to PA stuff above. Because that piece tries to resolve a common misconception born out of suckiness of ALSA documentation. Until ALSA gets a proper documentation we can refer to directly I would like the above piece about PA to stay there.

At the very least I want it to mention that you don't need PA for most sane things and refer to apulse for insane things. I'm ok with moving the snarky comments into the docstring for `apulse` package and halving the space it takes. Will that be agreeable?

7c6f434c commented:
> Eh? "nothing can update it"? What?
Once this file is generated, it is installed as `configuration.nix` on the user's system, and no automated tool ever touches it afterwards.

7c6f434c commented:
> I think we already had a wiki before and saw what that turned itself into.

For the record: it turned out that a MediaWiki with an open registration requires either too much skill or too much effort to prevent from becoming a spam dump (with spam even overwriting content). Nixos.Wiki seems to hope that a unique CAPTCHA will delay this problem, and in the worst case GitHub login can be made the only option.

makefu commented:
@7c6f434c you are only partly right. The first step to fight spam obviously is a technical measure (captchas), however the new wiki is watched over closely by  by a number of volunteers. There is a wiki-update bot in `#nixos-wiki` and we at least sight changes to ensure quality.

Some time ago someone asked how the archlinux wiki can stay such a wealthy resource and the answer was simple, there were people who cared. Now that the wiki is lead by a number of people and not a single person i think we can do something similar for the nixos community.

If you would like to help the effort you can join `nixos-wiki` on freenode!

7c6f434c commented:
@makefu well, you still need technical foundation to match — I mean, at some point I tried to roll back a spam attack on the old NixOS MediaWiki and suddenly understood that I just cannot understand how to untangle this sequence of history-masking-redirection-tricks. Some people did try to care, it just turned out that the people who cared at any given moment didn't have a combination of skill (and maybe MediaWiki access level?) to maintain defense efficiently.

Of course I do not deny the main effort of the team — creating and maintaining high-quality content. I just say that you have enough technical options to guarantee avoidance of the despair event horizon that the old wiki had.

(As for me, I am a person who can entertain people by describing how non-trivially my system is configured, and how it is not fully NixOS — so I can help people _debug_ via IRC, but writing useful Wiki content without immediate feedback would be hard)

edolstra commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  # or
+  # services.xserver.videoDrivers = [ "nvidiaLegacy304" ];
+  # or
+  # services.xserver.videoDrivers = [ "nvidiaLegacy173" ];
```
All of this should be detected automatically and emitted in `hardware-configuration.nix`.

oxij replied:
Right. As discussed above.

edolstra commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  #   # but you probably don't want to give it direct DRI access
+  #   # since OpenGL video drivers are pretty buggy and giving DRI
+  #   # access to WebGL might not be the best of ideas.
+  #   #          ++ [ "video" ];
```
This instruction should be absolutely unnecessary: logged-in users should automatically have access to sound and video devices via loginctl / udev rules. If not, that's a bug that should be fixed.

7c6f434c replied:
And how _not_ to give DRI access to a user even when logged in?

oxij replied:
> should automatically have access to sound and video devices via loginctl / udev rules

So you have a machine with a bunch of ssh accounts shared by a whole organization. Everyone should have access to sound and video devices by default? What about the "input" group that allows you to emulate key presses on evdev devices? I absolutely disagree.

wmertens replied:
@oxij in that case the loginctl/udev rules would definitely need adjusting. Only logging in on the console should give you these permissions.

edolstra commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -595,41 +655,172 @@ sub multiLineList {
+  #     umask 077
+  #
+  # into your ~/.xsession file. You should probably put the above
+  # xhost line there too.
```
This absolutely does not belong in configuration.nix.

oxij replied:
As noted above (https://github.com/NixOS/nixpkgs/pull/36274#issuecomment-371834510), I disagree, but fine, if we leave the user setup and then point to the manual for the rest, I would be ok with that.

oxij committed:
nixos: nixos-generate-config: document all the common things

oxij commented:
Updated.

Most frequently changing things are now gone to the configuration.nix(5). I left some of the DE setup in because xfce and slim are the most stable pieces of DE software ever, they are not going to change frequently (unlike Gnome and KDE). If you want to remove those lines I suggest we remove all DE setup then starting with KDE, since it's the most frequently changing thing in this config.

Unfree examples use "hello-unfree" package and so they are very un-corporate now.

I think I answered to all other nitpicks, feel free to reiterate if you disagree with the current state of the file.

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.36274&attempt_id=44e9ed88-4e9a-415d-9aa5-7ba20a479682)

The following builds were skipped because they don't evaluate on aarch64-linux: hello-unfree

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.36274&attempt_id=e9fb578d-7b49-47d5-b057-70c7aaeee2cb)

The following builds were skipped because they don't evaluate on x86_64-linux: hello-unfree

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

oxij commented:
A tiny piece of this was already merged in 92265e807e60fac8bbf3c08a82715b575b5e4e7e, I split other weakly-related parts into #41763 and #41764.

Ping!

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.36274&attempt_id=30005e11-3152-491b-9714-ce33379288e0)

The following builds were skipped because they don't evaluate on x86_64-linux: hello-unfree

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.36274&attempt_id=9821c008-f4f2-4e0e-a6c0-6ed3a8e28200)

The following builds were skipped because they don't evaluate on aarch64-linux: hello-unfree

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.36274&attempt_id=62790f2b-b316-4744-88d0-739f0813808d)

The following builds were skipped because they don't evaluate on aarch64-linux: hello-unfree

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.36274&attempt_id=eb58b618-18ad-46a3-9f7d-2ecf859560e5)

The following builds were skipped because they don't evaluate on x86_64-linux: hello-unfree

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

pasqui23 commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -605,35 +659,143 @@ sub multiLineList {
+  # \$ sudo -Hiu network firefox &disown
+  # \$ sudo -Hiu network qtox &disown
+  #
+  # from the account logged in into X11.
```
Honestly this feels like it should be a whole new option,which will add the required wrappers and sudo configurations.
Also for desktop users you should probably point to `programs.firejail`

wmertens commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -605,35 +659,143 @@ sub multiLineList {
+  # services.xserver.displayManager.auto = {
+  #   enable = true;
+  #   user = "me";
+  # };
```
Maybe this whole section could be marked as "for desktop use only"? Move the user stuff below it though.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/128

Mic92 commented:
There some useful bits in here that could be cherry-picked into our configuration. However given that the author is no longer working on this pull request, I will close this one.
